### ISSUE TITLE
Check estimator interface with duck typing rather than scikit-learn Mixin class inheritance

### ISSUE THREAD

**Issue by beckernick** [EpistasisLab/tpot#1106]
### General Introduction and Relevance

Currently, TPOT requires that estimators explicitly inherit from Scikit-learn Mixin classes in order to determine the nature of an estimator operator within the TPOTOperatorClassFactory and StackingEstimator. This Scikit-learn inheritance based programming model provides consistency, but also limits flexibility. Switching these checks to be duck-typing based rather than inheritance based would still preserve consistency but also allow users to use other libraries with TPOT, such as [cuML](https://github.com/rapidsai/cuml/). Using cuML with TPOT can be significantly faster, as shown in the example below that achieves a roughly 50x speedup with the example `config_dict`.

Note: Scikit-learn has moved to a duck typing based approach to determining the nature of whether an estimator is a classifier or regressor ([is_classifier](https://scikit-learn.org/stable/modules/generated/sklearn.base.is_classifier.html) and [is_regressor](https://scikit-learn.org/stable/modules/generated/sklearn.base.is_regressor.html)), which would make introducing this kind of change to the TPOT codebase fairly smooth and small.

### Additional Context

Currently, TPOT requires that estimators used with TPOT explicitly inherit from Scikit-learn Mixin classes in order to correctly function with the TPOTOperatorClassFactory. This Scikit-learn inheritance based programming model provides consistency, but also limits flexibility. A key limitation is that users cannot cleanly integrate estimators from libraries that do not directly depend on scikit-learn, but enforce the same API contract.

To resolve this problem more broadly, the PyData community has devoted and is continuing to devote significant effort to expand the use of a [duck typing](https://en.wikipedia.org/wiki/Duck_typing) based programming model. Duck typing in PyData is becoming more powerful as increasingly more libraries are standardizing on numpy and scikit-learn like API contracts rather than inheritance (Dask, xarray, Pangeo, CuPy, RAPIDS, and more). NumPy Enhancement Proposals [18](https://numpy.org/neps/nep-0018-array-function-protocol.html), [30](https://numpy.org/neps/nep-0030-duck-array-protocol.html), [35](https://numpy.org/neps/nep-0035-array-creation-dispatch-with-array-function.html), and [37](https://numpy.org/neps/nep-0037-array-module.html) describe a small portion of the already completed and ongoing work related to duck typing and dispatching in the community.

### Proposal

- In TPOTOperatorClassFactory, convert the Mixin class checks [1](https://github.com/EpistasisLab/tpot/blob/219f8c5abe43996abb2c19d6a1767083304a23d3/tpot/operator_utils.py#L181-L190) [2](https://github.com/EpistasisLab/tpot/blob/219f8c5abe43996abb2c19d6a1767083304a23d3/tpot/operator_utils.py#L294-L296) to instead check key attributes that determine whether an object is a certain type of estimator
    - Instead of ClassifierMixin, use is_classifier (or check for score and _estimator_type == “classifier”)
    - Instead of RegressorMixin, use is_regressor (or check for score and _estimator_type == “regressor”)
    - Instead of TransformerMixin, check for fit_transform
    - Instead of SelectorMixin, check for the key attributes that determine whether an object is SelectorMixin-like
        - Get_support
        - Transform
        - Inverse_transform
        - Fit_transform
- In the [transform method](https://github.com/EpistasisLab/tpot/blob/aea42a59b52ff9704f9f677a01ebb71fda4520cc/tpot/builtins/stacking_estimator.py#L86) of StackingEstimator, use is_classifier instead of ClassifierMixin subclass to check whether an estimator is Classifier-like
- Leave the BaseEstimator and Kernel checks in the export sub-function unchanged
    - Exporting a pipeline currently checks for subclassing of sklearn’s BaseEstimator and Kernel (the gaussian process kernel base class) as well as the Mixins. Because these checks are used with “or” logic, they do not block exporting pipelines including estimators that do not subclass sklearn’s base classes.

If there is potential interest, I'd be happy to open up a pull request for further discussion.

### Potential Impact

Adopting duck typing for these checks would immediately allow using cuML (a GPU-accelerated machine learning library with a scikit-learn-like API) with TPOT via a passed configuration dictionary. It would also open the door to other libraries (current and future), without requiring TPOT to have any knowledge or involvement.

As a concrete example of the benefit, using cuML estimators on the GPU instead of scikit-learn can provide a large speedup. Faster estimators can make a significant difference in the total runtime. The benchmark graph below provides a small example (reproducing gist provided at the bottom) which corresponds to a 49x speedup using cuML on a GPU vs the default scikit-learn with all available CPU cores (2.8 hours vs 3.5 minutes / 10206 seconds vs 208 seconds).

![tpot-cuml-speedup (2)](https://user-images.githubusercontent.com/8457388/90571497-dcd0f780-e17f-11ea-8d86-342b9b310249.png)

```
TPOT Estimator/Hyper-Parameter Configuration Dictionary
{"cuml.neighbors.KNeighborsClassifier": {"n_neighbors": [5, 15, 25]}, "cuml.svm.SVC": {"C": [0.001, 0.1, 1.0, 10.0]}, "cuml.ensemble.RandomForestClassifier": {"n_estimators": [100]}, "cuml.linear_model.LogisticRegression": {"C": [0.001, 0.1, 1.0, 100.0]}}
Data Shape: 300000 x 20 (`scikit-learn.datasets.make_classification`)
Generations:2
Pop Size:10
CV:3
```

Hardware Specs for the Loose Benchmark:
Dual Intel Xeon Platinum 8168, 2.7 GHz, 24-cores
NVIDIA V100 32 GB GPU

https://gist.github.com/beckernick/a2ce52850f17d694d31d3ac520ff3428

**Comment by weixuanfu** [comment]
I think it is very good idea. Please open a PR with a demo for using TPOT with a cuML configuration similar to TPOT's [default configurations](https://epistasislab.github.io/tpot/using/#built-in-tpot-configurations).

**Comment by beckernick** [comment]
Sounds good. Please feel free to assign this issue to me, if useful.

I'll open up a PR for further discussion/iteration.

**Comment by beckernick** [comment]
Closing as resolved by #1108
